ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,217 @@
+            parser_batch_size = self.parser_batch_size,
+            parser_task_count = self.parser_task_count,
+            "Preparing to fetch transactions"
+        );
+
+        let mut starting_version = self.current_version;
+        let mut num_fetches = 0;
+        let mut batches = vec![];
+
+        while num_fetches < self.parser_task_count && starting_version <= self.highest_known_version
+        {
+            let num_transactions_to_fetch = std::cmp::min(
+                self.parser_batch_size as u64,
+                self.highest_known_version - starting_version + 1,
+            ) as u16;
+
+            batches.push(TransactionBatchInfo {
+                start_version: starting_version,
+                num_transactions_to_fetch,
+            });
+            starting_version += num_transactions_to_fetch as u64;
+            num_fetches += 1;
+        }
+        batches
+    }
+
+    async fn fetch_raw_txns_with_retries(
+        context: Arc<Context>,
+        ledger_version: u64,
+        batch: TransactionBatchInfo,
+    ) -> Vec<TransactionOnChainData> {
+        loop {
+            match context.get_transactions(
+                batch.start_version,
+                batch.num_transactions_to_fetch,
+                ledger_version,
+            ) {
+                Ok(raw_txns) => return raw_txns,
+                Err(_err) => {
+                    continue;
+                },
+            }
+        }
+    }
+
+    /// Parses the table information from the raw transactions before converting to the api transactions,
+    /// optionally backup the rocksdb to gcs depending on epoch advancement or not.
+    async fn parse_table_info(
+        context: Arc<Context>,
+        raw_txns: Vec<TransactionOnChainData>,
+        db_writer: Arc<dyn DbWriter>,
+    ) -> Result<(), Error> {
+        if raw_txns.is_empty() {
+            return Ok(());
+        }
+
+        let start_millis = chrono::Utc::now().naive_utc();
+        let first_version = raw_txns.first().map(|txn| txn.version).unwrap();
+        let ledger_chain_id = context.chain_id().id();
+        let (_, _, block_event) = context
+            .db
+            .get_block_info_by_version(first_version)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Could not get block_info for start version {}",
+                    first_version,
+                )
+            });
+        let block_event_epoch = block_event.epoch();
+
+        let write_sets: Vec<WriteSet> = raw_txns.iter().map(|txn| txn.changes.clone()).collect();
+        let write_sets_slice: Vec<&WriteSet> = write_sets.iter().collect();
+        db_writer
+            .clone()
+            .index(
+                context.db.clone(),
+                first_version,
+                &write_sets_slice,
+                block_event_epoch,
+            )
+            .expect("Failed to process write sets and index to the table info rocksdb");
```

Comment:
probably we also want to log the duration of the db write and the duration of the write set parsing separately.